app-core/src/main/kotlin/kr/flab/wiki/core/domain/document/impl/PostDocumentUseCaseImpl.kt

```
@@ -0,0 +1,32 @@
+package kr.flab.wiki.core.domain.document.impl
+
+import kr.flab.wiki.core.common.exception.document.DocumentConflictException
+import kr.flab.wiki.core.domain.document.Document
+import kr.flab.wiki.core.domain.document.DocumentPostResult
+import kr.flab.wiki.core.domain.document.DocumentService
+import kr.flab.wiki.core.domain.document.usecases.PostDocumentUseCase
+
+class PostDocumentUseCaseImpl(
+    private val documentService: DocumentService
+) : PostDocumentUseCase {
+    override fun postDocument(document: Document): DocumentPostResult {
+        var serverDocument = documentService.getDocumentByTitle(document.title)
+        if (serverDocument.version != document.version) {
+            return DocumentPostResultImpl(document, serverDocument)
+        } else {
+            return try {
+                val updatedDocument = documentService.saveDocument(
+                    document.title,
+                    document.body,
+                    document.lastContributor,
+                    document.version
+                )
+                DocumentPostResultImpl(document, updatedDocument)
+            } catch (e: DocumentConflictException) {
+                println(e)
```

이런건 주의해주세요. 참고로 println 은 실행 성능을 꽤 잡아먹습니다. 왜 그런지는 https://stackoverflow.com/questions/18584809/java-system-out-effect-on-performance 한번 찬찬히 읽어보시고 저한테 설명해주세요.

---

그리고 이런경우 log 를 구현할 때 SLF4J 같은걸 이용해서,

```
logger.log("There were {} error found", e) 
```

같은 형태로 구현하는데,

```
logger.log("There were ${e} error found") 
```

로 구현하지 않는 이유에 대해서도 한번 생각해 보시면 좋겠네요.

참고로 이거 대답할 수 있으면 Java 10년차로 인정받으실 수 있는 고급 주제입니다.

---

아직 로그 관련쪽은 생각칠 못해서 detekt의 exception swallowing을 피하기 위해 println을 사용했었습니다. 로그 세팅도 알아보고 적용하도록 하겠습니다.